hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java

```diff
@@ -0,0 +1,171 @@
+package com.hazelcast.spi.impl;
+
+import com.hazelcast.instance.Node;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.util.ConcurrencyUtil;
+import com.hazelcast.util.ConstructorFunction;
+
+import java.util.Collection;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Monitors the operation threads to find and collect data of long running operations.
+ */
+final class SlowOperationDetector extends Thread {
```

Comment:
Design tip; don't extend from a thread for these things, but have an internal thread. It is an implementation detail and you don't want to leak it to the outside.

Especially when you have a single scanning thread, you can do something like this:

```
class SlowOperationDetector{

	void scan(){
		... actual scanning logic	
	}

	private class ScanThread extends Thread{
		public void run(){
			while(!shutdown){
				scan();			
			}		
		}	
	}
}
```
Now when you want to do a unit test, you don't call the start method but manually call the scan. You remember the story I told you about the axle? The scan method is the axle, and the thread is the engine that rotates that axle.